warp10/src/main/java/io/warp10/script/functions/ADDDURATION.java

```diff
@@ -0,0 +1,150 @@
+//
+//   Copyright 2020  SenX S.A.S.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.script.functions;
+
+import io.warp10.continuum.gts.UnsafeString;
+import io.warp10.continuum.store.Constants;
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptLib;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.MutablePeriod;
+import org.joda.time.ReadWritablePeriod;
+import org.joda.time.format.ISOPeriodFormat;
+
+import java.util.List;
+import java.util.Locale;
+
+public class ADDDURATION extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  public ADDDURATION(String name) {
+    super(name);
+  }
+  final private static WarpScriptStackFunction TSELEMENTS = new TSELEMENTS(WarpScriptLib.TSELEMENTS);
+  final private static WarpScriptStackFunction FROMTSELEMENTS = new FROMTSELEMENTS(WarpScriptLib.TSELEMENTSTO);
+
+  @Override
+  public WarpScriptStack apply(WarpScriptStack stack) throws WarpScriptException {
+
+    //
+    // Retrieve arguments
+    //
+
+    Object top = stack.pop();
+
+    if (!(top instanceof String || top instanceof Long)) {
+      throw new WarpScriptException(getName() + " expects an ISO8601 duration (a string) on top of the stack (see http://en.wikipedia.org/wiki/ISO_8601#Durations), or a number of duration (a Long).");
+    }
+
+    String duration;
+    long N = 1;
+    if (top instanceof String) {
+      duration = top.toString();
+    } else {
+      N = (Long) top;
+      top = stack.pop();
+      if (!(top instanceof String)) {
+        throw new WarpScriptException(getName() + " expects an ISO8601 duration (a string) in the second level of the stack (see http://en.wikipedia.org/wiki/ISO_8601#Durations).");
+      }
+      duration = top.toString();
+    }
+
+    String tz = null;
+    if (stack.peek() instanceof String) {
+      tz = stack.pop().toString();
+      if (!(stack.peek() instanceof Long)) {
+        throw new WarpScriptException(getName() + " operates on a tselements list, timestamp, or timestamp and timezone.");
+      }
+    } else if (!(stack.peek() instanceof List) && !(stack.peek() instanceof Long)) {
```

Comment:
For coherency you should either use !(condA) && !(CondB) or as before !(CondA || CondB)